prefer-template autofixer should not convert a string to a template if it contains an octal escape sequence

<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).

    This template is for bug reports. If you are here for another reason, please see below:

    1. To propose a new rule: https://eslint.org/docs/developer-guide/contributing/new-rules
    2. To request a rule change: https://eslint.org/docs/developer-guide/contributing/rule-changes
    3. To request a change that is not a bug fix, rule change, or new rule: https://eslint.org/docs/developer-guide/contributing/changes
    4. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

    Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

**Tell us about your environment**

* **ESLint Version:** 4.18.1
* **Node Version:** 9.6.1
* **npm Version:** 5.6.0

**What parser (default, Babel-ESLint, etc.) are you using?**

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

<!-- Paste your configuration below: -->
```yml
rules:
  prefer-template: error
parserOptions:
  ecmaVersion: 6
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

<!-- Paste the source code below: -->
```js
'\033' + foo
```

<!-- Paste the command you used to run ESLint: -->
```bash
eslint theFile.js --fix
```

**What did you expect to happen?**

I expected no autofix to happen, because the string `'\033'` contains an octal escape sequence, which is a syntax error when it appears in a template literal.

**What actually happened? Please include the actual, raw output from ESLint.**

The problem was automatically fixed, resulting in a syntax error.

```
1:2  error  Parsing error: Octal literal in strict mode
```

(Note: the error message is a bit misleading; this is a parsing error because the resulting string is a template literal, not because it's in strict mode.)

Hello,
I wanted to work on eslint and this seemed like a 'good first issue' to pick up. 

**My findings:**
I dug a bit deeper and found that the parser used by eslint(espree using acorn) is throwing this error in strict mode.

Specifically the error is thrown from here https://github.com/acornjs/acorn/blob/0d20f67b03e0784bdd18eb88e0f76e6a53043404/src/tokenize.js#L631

As for the fix there, this will not occur if you disable the 'strict' rule.

This is the file that i tried to replicate this issue
```js
"use strict";
function testIssue10031() {
    return "\033" + "postfix";
}
```
This threw me the error
```bash
  4:13  error  Parsing error: Octal literal in strict mode
```

If i disable the 'strict' rule, i get the proper rule suggestion for it
```js
/* eslint strict: 0 */
function testIssue10031() {
    return "\033" + "postfix";
}
```
```bash
  2:1   error  Missing JSDoc comment                          require-jsdoc
  2:10  error  'testIssue10031' is defined but never used     no-unused-vars
  3:12  error  Don't use octal: '\033'. Use '\u....' instead  no-octal-escape
  3:19  error  Unexpected string concatenation of literals    no-useless-concat
```

As eslint, i gather reports parse errors as also eslint errors, how should i proceed here? Should this be considered as a bug to be fixed?

--- a/lib/rules/prefer-template.js
+++ b/lib/rules/prefer-template.js
@@ -52,7 +52,7 @@ function isOctalEscapeSequence(node) {
         return false;
     }
 
-    const match = node.raw.match(/^([^\\]|\\[^0-7])*\\([0-3][0-7]{1,2}|[4-7][0-7]|[0-7])/);
+    const match = node.raw.match(/^([^\\]|\\([^0-7]|0[^0-7]))*\\([0-7]{1,3})/);
 
     if (match) {
 

--- a/lib/rules/prefer-template.js
+++ b/lib/rules/prefer-template.js
@@ -52,7 +52,7 @@ function isOctalEscapeSequence(node) {
         return false;
     }
 
-    const match = node.raw.match(/^([^\\]|\\([^0-7]|0[^0-7]))*\\([0-7]{1,3})/);
+    const match = node.raw.match(/^([^\\]|\\[^0-7])*\\([0-7]{1,3})/);
 
     if (match) {
 

--- a/lib/rules/prefer-template.js
+++ b/lib/rules/prefer-template.js
@@ -39,6 +39,45 @@ function getTopConcatBinaryExpression(node) {
 }
 
 /**
+ * Determines whether a given node is a octal escape sequence
+ * @param {ASTNode} node A node to check
+ * @returns {boolean} `true` if the node is an octal escape sequence
+ */
+function isOctalEscapeSequence(node) {
+
+    // No need to check TemplateLiterals – would throw error with octal escape
+    const isStringLiteral = node.type === "Literal" && typeof node.value === "string";
+
+    if (!isStringLiteral) {
+        return false;
+    }
+
+    const match = node.raw.match(/^([^\\]|\\[^0-7])*\\([0-3][0-7]{1,2}|[4-7][0-7]|[0-7])/);
+
+    if (match) {
+
+        // \0 is actually not considered an octal
+        if (match[2] !== "0" || typeof match[3] !== "undefined") {
+            return true;
+        }
+    }
+    return false;
+}
+
+/**
+ * Checks whether or not a node contains a octal escape sequence
+ * @param {ASTNode} node A node to check
+ * @returns {boolean} `true` if the node contains an octal escape sequence
+ */
+function hasOctalEscapeSequence(node) {
+    if (isConcatenation(node)) {
+        return hasOctalEscapeSequence(node.left) || hasOctalEscapeSequence(node.right);
+    }
+
+    return isOctalEscapeSequence(node);
+}
+
+/**
  * Checks whether or not a given binary expression has string literals.
  * @param {ASTNode} node - A node to check.
  * @returns {boolean} `true` if the node has string literals.
@@ -194,6 +233,22 @@ module.exports = {
         }
 
         /**
+         * Returns a fixer object that converts a non-string binary expression to a template literal
+         * @param {SourceCodeFixer} fixer The fixer object
+         * @param {ASTNode} node A node that should be converted to a template literal
+         * @returns {Object} A fix for this binary expression
+         */
+        function fixNonStringBinaryExpression(fixer, node) {
+            const topBinaryExpr = getTopConcatBinaryExpression(node.parent);
+
+            if (hasOctalEscapeSequence(topBinaryExpr)) {
+                return null;
+            }
+
+            return fixer.replaceText(topBinaryExpr, getTemplateLiteral(topBinaryExpr, null, null));
+        }
+
+        /**
          * Reports if a given node is string concatenation with non string literals.
          *
          * @param {ASTNode} node - A node to check.
@@ -216,9 +271,7 @@ module.exports = {
                 context.report({
                     node: topBinaryExpr,
                     message: "Unexpected string concatenation.",
-                    fix(fixer) {
-                        return fixer.replaceText(topBinaryExpr, getTemplateLiteral(topBinaryExpr, null, null));
-                    }
+                    fix: fixer => fixNonStringBinaryExpression(fixer, node)
                 });
             }
         }

--- a/tests/lib/rules/prefer-template.js
+++ b/tests/lib/rules/prefer-template.js
@@ -34,7 +34,8 @@ ruleTester.run("prefer-template", rule, {
 
         // https://github.com/eslint/eslint/issues/3507
         "var foo = `foo` + `bar` + \"hoge\";",
-        "var foo = `foo` +\n    `bar` +\n    \"hoge\";"
+        "var foo = `foo` +\n    `bar` +\n    \"hoge\";",
+        "var foo = `foo` + `\\0`;"
     ],
     invalid: [
         {

--- a/tests/lib/rules/prefer-template.js
+++ b/tests/lib/rules/prefer-template.js
@@ -187,6 +187,11 @@ ruleTester.run("prefer-template", rule, {
             code: "foo + 'handles unicode escapes correctly: \\x27'", // "\x27" === "'"
             output: "`${foo  }handles unicode escapes correctly: \\x27`",
             errors
+        },
+        {
+            code: "foo + 'does not autofix octal escape sequence' + '\\033'",
+            output: null,
+            errors
         }
     ]
 });

--- a/tests/lib/rules/prefer-template.js
+++ b/tests/lib/rules/prefer-template.js
@@ -26,6 +26,7 @@ const ruleTester = new RuleTester({ parserOptions: { ecmaVersion: 6 } });
 ruleTester.run("prefer-template", rule, {
     valid: [
         "'use strict';",
+        "var foo = 'foo' + '\\0';",
         "var foo = 'bar';",
         "var foo = 'bar' + 'baz';",
         "var foo = foo + +'100';",
@@ -34,8 +35,7 @@ ruleTester.run("prefer-template", rule, {
 
         // https://github.com/eslint/eslint/issues/3507
         "var foo = `foo` + `bar` + \"hoge\";",
-        "var foo = `foo` +\n    `bar` +\n    \"hoge\";",
-        "var foo = `foo` + `\\0`;"
+        "var foo = `foo` +\n    `bar` +\n    \"hoge\";"
     ],
     invalid: [
         {
@@ -205,8 +205,8 @@ ruleTester.run("prefer-template", rule, {
             errors
         },
         {
-            code: "foo + '\\0 other test \\033'",
-            output: null,
+            code: "foo + '\\0'",
+            output: "`${foo  }\\0`",
             errors
         }
     ]

--- a/tests/lib/rules/prefer-template.js
+++ b/tests/lib/rules/prefer-template.js
@@ -192,6 +192,16 @@ ruleTester.run("prefer-template", rule, {
             code: "foo + 'does not autofix octal escape sequence' + '\\033'",
             output: null,
             errors
+        },
+        {
+            code: "foo + '\\n other text \\033'",
+            output: null,
+            errors
+        },
+        {
+            code: "foo + '\\\\033'",
+            output: "`${foo  }\\\\033`",
+            errors
         }
     ]
 });

--- a/tests/lib/rules/prefer-template.js
+++ b/tests/lib/rules/prefer-template.js
@@ -202,6 +202,11 @@ ruleTester.run("prefer-template", rule, {
             code: "foo + '\\\\033'",
             output: "`${foo  }\\\\033`",
             errors
+        },
+        {
+            code: "foo + '\\0 other test \\033'",
+            output: null,
+            errors
         }
     ]
 });

